### PR TITLE
Use resource handles

### PR DESCRIPTION
**Pull request by AnonymousProgrammerLandau**
I think the library benefits from further use of resource handles (RAII).

 - [x] Encapsulate in a type the concept of calling exactly once to Lua for statements like `Res r = state["lua_function"](...)`.
 - [x] Use the destructor of a type to clear the stack (`lua_State`) after modifying activities.

### CODE REVIEW

File: include/selene/State.h [hunk 1]
```diff
@@ -134,9 +134,14 @@ class State {
...
-        bool result = !luaL_dostring(_l, code);
-        if(result) lua_settop(_l, 0);
-        return result;
+        ResetStackOnScopeExit savedStack(_l);
+        int status = luaL_dostring(_l, code);
+        if(status) {
+            _exception_handler->Handle_top_of_stack(status, _l);
+            return false;
+        }
+        savedStack.KeepChanges();
```
Comment (AnonymousProgrammerLandau) [inline include/selene/State.h]: The member functions `Load(...)`, `OpenLib(...)`, `Push(...)` and `operator()(...)` leave objects on the stack. There is no function to remove objects from the stack.
 1. Make them to not leave objects on the stack and remove `Push(...)` and `Read(...)`.
 2. Add `Pop(...)` or similar functions.
I see no real use of managing objects on the stack at this level of abstraction - apart from using Lua's implicit type conversions. The latter seems unimportant enough to me to favor item 1.
@jeremyong: Am I missing a use case?

**Comment by jeremyong** [comment]
Sorry it's taken me so long to look at this. Life has been busy!

Regarding your last question, yea some of those functions exist due to legacy (how they were implemented in an earlier version). It makes to remove the abstraction if it is not necessary.

One thing I've been thinking about is implementing the selector without the use of std::function (which strikes me as entirely doable), which makes me pause when accepting this PR. The functions we are storing in `std::function`s are not small enough to invoke the small functor optimization of some compilers I believe, which is causing expensive copying and slowing down code compilation a good deal as well.

**Comment by AnonymousProgrammerLandau** [comment]
I have no further contributions to this PR.

I also see no contradiction to #94. Do you have further points on that?

**Comment by jeremyong** [comment]
Sorry it took so long to get to this. Thanks again for the contribution!

**Comment by AnonymousProgrammerLandau** [comment]
> Sorry it took so long to get to this. Thanks again for the contribution!
For each commit you accept I don't need to worry about maintaining a patchset for. So: Thanks for accepting!

**Pull request merged**